File: R/probability_mass.R
#===============================================================================
# probability_mass.R
#===============================================================================

# Imports ======================================================================

#' @import VGAM




# Functions ====================================================================

#' @title Probability mass for a bbs in the independent case
#'
#' @param x vector giving the number of successes for each group.
#' @param size vector giving the number of trials for each group.
#' @param shape1,shape2 the two (positive) shape parameters of the standard
#'   beta distribution.
#' @return numeric, the probability mass.
probability_mass_independent <- function(
  x,
  size,
  prob = 0.5,
  rho = 0,
  shape1 = NULL,
  shape2 = NULL,
  ...
) {
  prod(
    sapply(
      1:length(x),
      function(index) {
        if (is.null(shape1) && is.null(shape2)) {
          dbetabinom(
            x[[index]],
            size[[index]],
            prob = prob[[index]],
            rho = rho[[index]],
            ...
          )
        } else if (is.numeric(shape1) && is.numeric(shape2)) {
          dbetabinom.ab(
            x[[index]],
            size[[index]],
            shape1 = shape1[[index]],
            shape2 = shape2[[index]],
            ...
          )
        } else if (is.numeric(shape1) && rho == 0) {
          function(index) {
            dbetabinom.ab(
              x[[index]],
              size[[index]],
              shape1 = shape1[[index]],
              shape2 = shape1[[index]]*(1-prob[[index]])/prob[[index]],
              ...
            )
          }
        } else {
          stop("Invalid arguments")
        }
      }
    )
  )
}

#' @title The integrand for computing probability mass in the dependent case
#'
#' @param t vector of values
#' @param x vector giving the number of successes for each group.
#' @param size vector giving the number of trials for each group.
#' @param shape1,shape2 the two (positive) shape parameters of the standard
#'   beta distribution. 
#' @return numeric, the value of the evaluated integrand.
integrand_dependent <- function(t, x, size, shape1, shape2) {
  sapply(
    t,
    function(t_i) {
      prod(
        sapply(
          1:length(x),
          function(i) {
            q = qbeta(t_i, shape1[[i]], shape2[[i]])
            if (x[[i]] == 0) {
              (1 - q)^(size[[i]])
            } else if (x[[i]] == size[[i]]) {
              q^(x[[i]])
            } else {
              q^x[[i]] * (1 - q)^(size[[i]] - x[[i]])
            }
          }
        )
      )
    }
  )
}

#' @title Logarithm of the integral component of probability mass
#'
#' @param x vector giving the number of successes for each group.
#' @param size vector giving the number of trials for each group.
#' @param shape1,shape2 the two (positive) shape parameters of the standard
#'   beta distribution. 
#' @return numeric, the value of the evaluated integral.
log_integral_dependent <- function(x, size, shape1, shape2) {
  integral <- tryCatch(
    integrate(
      integrand_dependent,
      0,
      1,
      x = x,
      size = size,
      shape1 = shape1,
      shape2 = shape2,
      rel.tol = 1e-15
    )[["value"]],
    error = function(e) NA
  )
  log(integral)
}

#' @title Logarithm of the coefficient component of probability mass
#'
#' @param x vector giving the number of successes for each group.
#' @param size vector giving the number of trials for each group.
#' @param shape1,shape2 the two (positive) shape parameters of the standard
#'   beta distribution. 
#' @return numeric, the value of the log-coefficient.
log_coefficient_dependent <- function(x, size) {
  sum(
    sapply(
      1:length(x),
      function(i) {
        log(choose(size[[i]], x[[i]]))
      }
    )
  )
}

#' @title Probability mass for a bbs in the dependent case
#'
#' @param x vector giving the number of successes for each group.
#' @param size vector giving the number of trials for each group.
#' @param shape1,shape2 the two (positive) shape parameters of the standard
#'   beta distribution. 
#' @return numeric, the probability mass.
probability_mass_dependent <- function(x, size, shape1, shape2) {
  exp(
    sum(
      log_coefficient_dependent(x, size),
      log_integral_dependent(x, size, shape1, shape2)
    )
  )
}

File: betabernsum/probability_mass.py
#===============================================================================
# probability_mass.py
#===============================================================================

# Imports ======================================================================

from functools import reduce
from math import log
from operator import mul
from scipy.integrate import quad
from scipy.misc import comb
from scipy.special import beta as beta_function
from scipy.stats import beta




# Functions ====================================================================

def prod(iterable):
    """"Product of values in an iterable
    
    Parameters
    ----------
    iterable
        an iterable
    
    Returns
    -------
    the product
    """
    return reduce(mul, iterable, 1)


def betabinom_pmf(k, n, a, b):
    """Probability mass function for a beta binomial distribution

    Parameters
    ----------
    k : int
        number of successes
    n : int
        number of trials
    a : float
        first (positive) shape parameter
    b : float
        second (positive) shape parameter

    Returns
    -------
    float
        the probability mass
    """

    return comb(n, k) * beta_function(k + a, n - k + b) / beta_function(a, b)


def probability_mass_independent(k, n, a, b):
    """Probability mass for a bbs in the independent case

    Parameters
    ----------
    k
        iterable giving the number of successes for each group
    n
        iterable giving the number of trials for each group
    a
        iterable giving the first shape parameter for each group
    b
        iterable giving the second shape parameter for each group
    
    Returns
    -------
    float
        the probability mass
    """
    
    assert len(k) == len(n) == len(a) == len(b)
    return prod(
        betabinom_pmf(k_i, n_i, a_i, b_i)
        for k_i, n_i, a_i, b_i in zip(k, n, a, b)
    )


def evaluate_factor(t, k, n, a, b):
    """Evaluate a factor for the product in `integrand_dependent`

    t : float
        a value on the domain of integration
    k : int
        a number of successes
    n : int
        a number of trials
    a : float
        a first shape parameter
    b : float
        a second shape parameter
    
    Returns
    -------
    float
        value of a factor
    """

    q = beta.ppf(t, a, b)
    if k == 0:
        return (1 - q)**n
    elif k == n:
        return q**k
    else:
        return q**k * (1 - q)**(n - k)
    

def integrand_dependent(t, k, n, a, b):
    """The integrand for computing probability mass in the dependent case

    Parameters
    ----------
    t
        iterable of values on the domain of integration
    k
        iterable giving the number of successes for each group
    n
        iterable giving the number of trials for each group
    a
        iterable giving the first shape parameter for each group
    b
        iterable giving the second shape parameter for each group

    Returns
    -------
    float
    """

    assert len(k) == len(n) == len(a) == len(b)
    return tuple(
        prod(
            evaluate_factor(t_i, k_j, n_j, a_j, b_j)
            for k_j, n_j, a_j, b_j in zip(k, n, a, b)
        )
        for t_i in t
    )


def log_integral_dependent(k, n, a, b):
    """Get logarithm of integral
    
    Parameters
    ----------
    k
    n
    a
    b
    """
    
    return log(quad(integrand_dependent, 0, 1, args = (k, n, a, b))[0])
